packages/firestore/src/core/query.ts

```diff
@@ -686,13 +644,86 @@ export class NanFilter extends Filter {
   toString(): string {
     return `${this.field.canonicalString()} IS NaN`;
   }
+}
 
-  isEqual(other: Filter): boolean {
-    if (other instanceof NanFilter) {
-      return this.field.isEqual(other.field);
-    } else {
-      return false;
-    }
+/**
+ * Filter that matches on key fields (i.e. '__name__').
+ */
+export class KeyFieldFilter extends FieldFilter {
+  constructor(field: FieldPath, op: Operator, value: RefValue) {
+    super(field, op, value);
+  }
+
+  matches(doc: Document): boolean {
+    const refValue = this.value as RefValue;
+    const comparison = DocumentKey.comparator(doc.key, refValue.key);
+    return this.matchesComparison(comparison);
+  }
+}
+
+/**
+ * A Filter that implements the array-contains operator. When a user specifies
+ * query.where('some.field', 'array-contains', 42), this filter will match
+ * documents that:
+ *
+ *   * contain a field at some.field;
+ *   * that field is an array; and
+ *   * that array contains the value 42.
+ */
```

Comment:
(you are 110% welcome to ignore, but...)  TBH these comments feel a little overkill and out-of-place here (they would be better suited for inclusion in our public API docs or something).  I'm also not wild about `*` bullet lists inside `/*` comments.  I also want to get this PR to be net negative lines of code (and these descriptions are longer than the actual implementation). 😛